python/cudf_polars/cudf_polars/experimental/expressions.py

```diff
@@ -0,0 +1,485 @@
+    return expr.evaluate(df, mapping=dict(zip(children, references, strict=True)))
+
+
+def evaluate_chunk_multi_agg(
+    df: DataFrame,
+    exprs: Sequence[Expr],
+    children: tuple[Expr, ...],
+    *references: Column,
+) -> tuple[Column, ...]:
+    """Evaluate multiple aggregations."""
+    return tuple(
+        expr.evaluate(df, mapping=dict(zip(children, references, strict=True)))
+        for expr in exprs
+    )
+
+
+def combine_chunks_multi_agg(
+    column_chunks: Sequence[tuple[Column]],
+    combine_aggs: Sequence[Agg],
+    finalize: tuple[plc.DataType, str] | None,
+    name: str,
+) -> Column:
+    """Aggregate Column chunks."""
+    column_chunk_lists = zip(*column_chunks, strict=True)
+
+    combined = [
+        agg.op(
+            Column(
+                plc.concatenate.concatenate([col.obj for col in column_chunk_list]),
+                name=column_chunk_list[0].name,
+            )
+        )
+        for agg, column_chunk_list in zip(combine_aggs, column_chunk_lists, strict=True)
+    ]
+
+    if finalize:
+        # Perform optional BinOp on combined columns
+        dt, op_name = finalize
+        op = getattr(plc.binaryop.BinaryOperator, op_name)
+        col = Column(plc.binaryop.binary_operation(*(c.obj for c in combined), op, dt))
+    else:
+        assert len(combined) == 1
+        col = combined[0]
+
+    return col.rename(name)
+
+
+def shuffle_child(
+    child: IR,
+    on: Expr,
+    named_expr: NamedExpr,
+    child_partition_info: PartitionInfo,
+    config_options: ConfigOptions,
+) -> tuple[IR, MutableMapping[Any, Any]]:
+    """
+    Shuffle the child IR on a single expression.
+
+    Parameters
+    ----------
+    child
+        Child IR we are shuffling.
+    on
+        Single Expr to shuffle on.
+    named_expr
+        Outer NamedExpr needing this shuffle.
+    child_partition_info
+        IR partitioning information.
+    config_options
+        GPU engine configuration options.
+
+    Returns
+    -------
+    shuffled
+        New (shuffled) child. The corresponding DataFrame
+        container will only contain the columns associated
+        with the shuffle keys (``on``).
+    graph
+        Task graph needed to shuffle child.
+    """
+    graph: MutableMapping[Any, Any] = {}
+
+    if child_partition_info.partitioned_on == (
+        named_expr.reconstruct(on),
+    ):  # pragma: no cover
+        # No shuffle necessary
+        return child, graph
+
+    child_partition_info = PartitionInfo(count=child_partition_info.count)
+    shuffle_on = (
+        named_expr.reconstruct(
+            FusedExpr(
+                on.dtype,
+                on,
+                *[c for c in named_expr.value.children if isinstance(c, FusedExpr)],
+            )
+        ),
+    )
+
+    pi: MutableMapping[IR, PartitionInfo] = {child: child_partition_info}
+    schema = {col.name: col.dtype for col in traversal([on]) if isinstance(col, Col)}
```

Comment:
This seems very wrong, it's extracting all columns that are leaves of `on` (an arbitrary expression that we are shuffling on) and using those as the schema of the `Select` call, which only selects a single column (`shuffle_on`).

What are you trying to do here?

I suppose what you're trying to do is determine which columns of the `child` are required to evaluate `shuffle_on`?

Reply:
I can avoid all this for now if you'd rather keep things simple (and there could definitely be a mistake).

The general idea is that we need to shuffle the input IR node to evaluate this expression, but the output is only a single column. So, in order to avoid shuffling a bunch of columns that we don't actually need to evaluate the expression, we (1) look for the underlying columns needed by the expression, (2) drop all other columns from the input IR node, and (3) shuffle only the columns we need.

Reply:
That makes sense, but those columns need to be in the list of expressions you're selecting

Reply:
>but those columns need to be in the list of expressions you're selecting

I don't think I understand this comment very well. Are you saying that we cannot drop columns because we might need it in the overarching `Select`, even if we don't need it for the current `FusedExpr` we are processing in this function?

When we drop columns and shuffle IR, the modified/shuffled IR is only used for this specific `FusedExpr`. If other `FusedExpr` nodes depended on other columns, they would still be referencing the original input IR.

Reply:
As I understand it, you're shuffling on some expression `A`. This expression depends on some columns `(B, C)`. The `child` input might have additional columns `(D, E, ...)`. And you want to ensure that you only move around the data you need to evaluate `A`.

So I think this is:

```
necessary_columns = [e for e in traversal([on]) if isinstance(e, expr.Col)]

input = Select({c.name: c.dtype for c in necessary_columns},
               necessary_columns,
               False,
               child)

shuffled = Shuffle(..., shuffle_on, input)
```
No?